FWCore/Framework/src/EventProcessor.cc

```diff
@@ -718,113 +716,148 @@ namespace edm {
 
     //make the services available
     ServiceRegistry::Operate operate(serviceToken_);
 
-    using namespace edm::waiting_task::chain;
+    if (beginJobSucceeded_) {
+      endStream(c);
+    }
 
-    oneapi::tbb::task_group group;
-    edm::FinalWaitingTask waitTask{group};
+    if (beginJobStartedModules_) {
+      try {
+        convertException::wrap([this, &c]() { schedule_->endJob(c); });
+      } catch (cms::Exception const& ex) {
+        c.addException(ex);
+      }
+      for (auto& subProcess : subProcesses_) {
+        try {
+          convertException::wrap([&subProcess, &c]() { subProcess.doEndJob(c); });
+        } catch (cms::Exception const& ex) {
+          c.addException(ex);
+        }
+      }
+      c.call(std::bind(&InputSource::doEndJob, input_.get()));
+      if (looper_) {
+        c.call(std::bind(&EDLooperBase::endOfJob, looper()));
+      }
+      if (c.hasThrown()) {
+        c.rethrow();
+      }
+    }
+  }
 
+  void EventProcessor::endStream(ExceptionCollector& collector) noexcept {
+    std::mutex collectorMutex;
+
+    // This will process streams concurrently, but not modules in the
+    // same stream or SubProcesses.
+    oneapi::tbb::task_group group;
+    FinalWaitingTask finalWaitingTask{group};
+    using namespace edm::waiting_task::chain;
     {
-      //handle endStream transitions
-      edm::WaitingTaskHolder taskHolder(group, &waitTask);
-      std::mutex collectorMutex;
+      WaitingTaskHolder taskHolder(group, &finalWaitingTask);
       for (unsigned int i = 0; i < preallocations_.numberOfStreams(); ++i) {
-        first([this, i, &c, &collectorMutex](auto nextTask) {
-          std::exception_ptr ep;
-          try {
+        first([this, i, &collector, &collectorMutex](auto nextTask) {
+          {
             ServiceRegistry::Operate operate(serviceToken_);
-            this->schedule_->endStream(i);
-          } catch (...) {
-            ep = std::current_exception();
-          }
-          if (ep) {
-            std::lock_guard<std::mutex> l(collectorMutex);
-            c.call([&ep]() { std::rethrow_exception(ep); });
-          }
-        }) | then([this, i, &c, &collectorMutex](auto nextTask) {
-          for (auto& subProcess : subProcesses_) {
-            first([this, i, &c, &collectorMutex, &subProcess](auto nextTask) {
-              std::exception_ptr ep;
-              try {
-                ServiceRegistry::Operate operate(serviceToken_);
-                subProcess.doEndStream(i);
-              } catch (...) {
-                ep = std::current_exception();
-              }
-              if (ep) {
-                std::lock_guard<std::mutex> l(collectorMutex);
-                c.call([&ep]() { std::rethrow_exception(ep); });
-              }
-            }) | lastTask(nextTask);
+            schedule_->endStream(i, collector, collectorMutex);
+            for (auto& subProcess : subProcesses_) {
+              subProcess.doEndStream(i, collector, collectorMutex);
```

Comment:
This might make sense in endStream. I guess I'm leaning against this because there isn't any concurrency in endJob and the mutex isn't needed in that part.  If you and @Dr15Jones want me to, I'll try to make those changes also.

Reply:
Ok. I was thinking the exception message collection is not performance critical, and therefore preferring thread-safety could make sense. We could leave this also to the next time the `ExceptionCollector` is used in a concurrent context.